### PR TITLE
[1.24.0 WIP] Fix path serialization

### PR DESCRIPTION
**Pull request by Karrenbelt**
## Proposed changes

Tests for config serialization containing paths. In furtherance of #359 

## Fixes

If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: tests/test_configurations/test_base.py [hunk 1]
```diff
@@ -1123,3 +1130,125 @@ def test_component_id_from_json():
...
+
+        def sorted_indices(li: List[str]) -> List[int]:
+            return sorted(range(len(li)), key=lambda i: li[i])
+
+        # since automagically converted to POSIX compliant-format when on Linux,
+        # hard-coded the expected posix order to ensure consistency across OS.
+        expected = [3, 2, 5, 0, 4, 1, 11, 9, 7, 10, 8, 6, 14, 12, 13, 15, 16, 17]
+        posix_paths = [Path(p).as_posix() for p in self.raw_paths]
+        assert sorted_indices(self.raw_paths) == sorted_indices(posix_paths)
+        assert sorted_indices(self.raw_paths) == expected
```
Comment (Karrenbelt) [inline tests/test_configurations/test_base.py:1197]: this should fail on Windows

File: tests/test_configurations/test_base.py [hunk 2]
```diff
@@ -1123,3 +1130,125 @@ def test_component_id_from_json():
...
+
+    def test_skill_component_configuration_serialization(self) -> None:
+        """Test SkillComponentConfiguration serialization"""
+
+        config = SkillComponentConfiguration(
+            class_name="class_name", file_path=self.raw_paths[0]
+        )
+        expected = "813554f9b5a750bad28bfd7005705bfc001003ac2ed2970d9030cf1b2d2ec37db69390cfc810f141e228bd3232238a9462259c2df6a3f87631606f47d0183cf8"
```
Comment (angrybayblade) [inline tests/test_configurations/test_base.py:1205]: Please make this a constant and add an explanation

File: tests/test_configurations/test_base.py [hunk 3]
```diff
@@ -1123,3 +1130,125 @@ def test_component_id_from_json():
...
+        name, author = "name", "author"
+        config = SkillConfig(
+            name=name,
+            author=author,
+            build_entrypoint=self.raw_paths[0],
+            build_directory=self.raw_paths[0],
+        )
+        expected = "6de1f8810c55f1ae3d9cea3c58bf41e45049ca8a8078540d5550246a0b3783dbed60672f559a38d51cf802416b8f49b7b649e367dc840ed01cda38ef19a57182"
+        assert self.yaml_config_dump_load_equal(config.json)
+        assert self.get_hexdigest_from_config_json(config) == expected
```
Comment (angrybayblade) [inline tests/test_configurations/test_base.py:1221]: You should also add more explicit check for serialised path to make sure they're dumped as posix paths

Comment (Karrenbelt) [reply tests/test_configurations/test_base.py:1221]: I can recursively check the values in the dictionary (json) and if string type check that converting to posix format they remain constant.

File: tests/test_configurations/test_base.py [hunk 4]
```diff
@@ -1123,3 +1131,138 @@ def test_component_id_from_json():
...
+
+class TestConfigurationContainingPathSerialization:
+    """Test configurations containing paths are deterministic across different OS"""
+
+    def setup(self):
+        """Setup test"""
+
+        # starting from posix-compliant format
+        self.raw_paths = [
+            "C:/Documents/Newsletters/Summer2018.pdf",
```
Comment (angrybayblade) [inline tests/test_configurations/test_base.py:1144]: Re your comment on removing the root, afaic we use paths which are relative to the project so I don't think that should be an issue

Comment (Karrenbelt) [reply tests/test_configurations/test_base.py:1144]: that would make sense if true but I'm not so sure. For example, we also store paths to private key files. In that is what we want to opt for though, I suggest we enforce it. We can change ticket #440 to address this, by introducing a check that ensures all paths are relative to specific root (e.g. the project)

**Comment by Karrenbelt** [comment]
failures
- macos 3.10 - FetchAI's testnet not working properly
    ```
    FAILED tests/test_cli/test_transfer.py::TestCliTransferFetchAINetworkWithPassword::test_integration
    FAILED tests/test_cli/test_transfer.py::TestCliTransferFetchAINetwork::test_integration
    ```
- windows 3.8 - not seen before, expect due to relatively short timeout
    ```
    tick_interval = 0.1
    wait_for_condition(lambda: behaviour.counter >= 1, timeout=tick_interval * 2)
    ...
    FAILED tests/test_agent_loop.py::TestAsyncAgentLoop::test_behaviour_act
    ```

**Review by angrybayblade — APPROVED** [review]

**Pull request merged**